File: 0501-1000/0588. Design In-Memory File System.go
package main
// Hard
// https://leetcode.cn/problems/design-in-memory-file-system/

import (
	"sort"
	"strings"
)
type FileSystem struct {
    root *Node
}

type Node struct {
    isFile bool
    content string
    children map[string]*Node
}

func Constructor() FileSystem {
    return FileSystem{
        root: &Node{
            children: make(map[string]*Node),
        },
    }
}

func (this *FileSystem) Ls(path string) []string {
    node := this.traverse(path)

    if node.isFile {
        parts := splitPath(path)
        return []string{parts[len(parts) - 1]}
    }

    res := []string{}
    for name := range node.children {
        res = append(res, name)
    }
    sort.Strings(res)
    return res
}

func (this *FileSystem) Mkdir(path string)  {
    parts := splitPath(path)
    cur := this.root

    for _, part := range parts {
        if cur.children[part] == nil {
            cur.children[part] = &Node{
                children: make(map[string]*Node),
            }
        }
        cur = cur.children[part]
    }
}

func (this *FileSystem) AddContentToFile(filePath string, content string)  {
    parts := splitPath(filePath)
    cur := this.root

    for i := 0; i < len(parts) - 1; i++ {
        if cur.children[parts[i]] == nil {
            cur.children[parts[i]] = &Node{
                children: make(map[string]*Node),
            }
        }
        cur = cur.children[parts[i]]
    }
    
    fileName := parts[len(parts) - 1]
    if cur.children[fileName] == nil {
        cur.children[fileName] = &Node{
            isFile: true,
            content: content,
            children: make(map[string]*Node),
        }
    } else {
        cur.children[fileName].content += content
    }
}

func (this *FileSystem) ReadContentFromFile(filePath string) string {
    node := this.traverse(filePath)
    return node.content
}

func (this *FileSystem) traverse(path string) *Node {
    if path == "/" {
        return this.root
    }

    parts := splitPath(path)
    cur := this.root

    for _, part := range parts {
        cur = cur.children[part]
    }
    return cur
}

func splitPath(path string) []string {
    if path == "/" {
        return []string{}
    }
    
    parts := []string{}
    for _, p := range strings.Split(path, "/") {
        if p != "" {
            parts = append(parts, p)
        }
    }
    return parts
}

File: 0501-1000/0588. Design In-Memory File System.py
# Time:O(L) L = path length
# Space:O(N) N = node number

from typing import List

class TrieNode:
    def __init__(self):
        self.children = {}
        self.isFile = False # directory/file
        self.content = ""

class FileSystem:

    def __init__(self):
        self.root = TrieNode() 
    
    def _split(self, path):
        if path == "/":
            return []
        return path.split('/')[1:]

    def ls(self, path: str) -> List[str]:
        node = self.root 
        parts = self._split(path)
        for p in parts:
            if p not in node.children:
                return []
            node = node.children[p]
        
        if node.isFile:
            return [parts[-1]]
        else:
            return sorted(node.children.keys())

    def mkdir(self, path: str) -> None:
        node = self.root
        parts = self._split(path)
        for p in parts:
            if p not in node.children:
                node.children[p] = TrieNode()
            node = node.children[p]

    def addContentToFile(self, filePath: str, content: str) -> None:
        node = self.root
        parts = self._split(filePath)
        for p in parts[:-1]:
            if p not in node.children:
                node.children[p] = TrieNode()
            node = node.children[p]
        
        fileName = parts[-1]
        if fileName not in node.children:
            node.children[fileName] = TrieNode()
        fileNode = node.children[fileName]
        fileNode.isFile = True
        fileNode.content += content

    def readContentFromFile(self, filePath: str) -> str:
        node = self.root
        parts = self._split(filePath)
        for p in parts:
            node = node.children[p]
        return node.content
    